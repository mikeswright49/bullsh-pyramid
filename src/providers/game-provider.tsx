import React, { useEffect, useState } from 'react';
import { GameState } from 'types/game-state';
import { GameStore } from 'src/stores/game-store';
import { GameStage } from 'src/enums/game-stage';

export const DEFAULT_GAME_STATE: GameState = {
    id: '',
    gameStage: GameStage.Initiation,
    activeRow: 0,
    activeIndex: 0,
    cardCount: 4,
    flipDelay: 5000,
    players: [],
    tiers: [],
};
export const GameContext = React.createContext<GameState>(DEFAULT_GAME_STATE);

export function GameProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const { gameId } = props;
    const [gameState, setGameState] = useState<GameState>(DEFAULT_GAME_STATE);

    useEffect(() => {
        if (!gameId) {
            return;
        }

        GameStore.subscribeToGame(gameId, (updatedGameState) => {
            setGameState(updatedGameState);
        });

        return () => GameStore.unsubscribeToGame();
    }, [gameId]);

    return <GameContext.Provider value={gameState}>{props.children}</GameContext.Provider>;
}
