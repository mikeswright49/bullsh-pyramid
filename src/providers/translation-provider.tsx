import React, { useContext } from 'react';
import { useEffect, useState } from 'react';
import { GameContext } from './game-provider';
import { TRANSLATIONS } from '../translations/translations';

const INTERPOLATE_REGEX = /{{.*?}}/gi;
const INTERPOLATE_PROPERTY_REGEX = /{{(.*)?}}/i;
const INTERPOLATE_CAPTURE_COUNT = 2;

export type Translations = {
    translations: { [key: string]: string };
    translate: (key: string, variables?: { [key: string]: string | number | boolean }) => string;
};

export const TranslationContext = React.createContext<Translations>({
    translate: (a) => a,
    translations: {},
});

export function TranslationProvider(props: React.PropsWithChildren<unknown>) {
    const { language } = useContext(GameContext);
    const [translations, setTranslations] = useState({});
    useEffect(() => {
        setTranslations(TRANSLATIONS[language || 'en']);
    }, [language]);

    /**
     * Returns the nested property value (userMetadata.giftcard.bacon)
     * @param source property value
     */
    function getProperty(source: string[], parent: unknown): string {
        if (!source || !parent) {
            return;
        }
        if (source.length > 1) {
            const property: string = source.shift();
            return this.getProperty(source, parent[property.trim()]);
        }
        return parent[source[0].trim()];
    }

    /**
     * Evaluates the {{}} interpolated strings
     * @param source Source content to be parsed
     * @param data Data object to read values off of
     */
    function evaluateInterpolate(
        source: string,
        data: { [key: string]: string | number | boolean }
    ): string {
        if (!data) {
            return source;
        }

        const matches = source.match(INTERPOLATE_REGEX) || [];
        matches.forEach((match: string) => {
            const properties = INTERPOLATE_PROPERTY_REGEX.exec(match);
            if (properties && properties.length === INTERPOLATE_CAPTURE_COUNT) {
                const property = getProperty(properties[1].split('.'), data);
                source = source.replace(match, property || '');
            }
        });
        return source;
    }

    function translate(key: string, variables?: { [key: string]: string | number | boolean }) {
        if (!translations) {
            return '';
        }
        return evaluateInterpolate(translations[key], variables);
    }

    return (
        <TranslationContext.Provider value={{ translations, translate }}>
            {props.children}
        </TranslationContext.Provider>
    );
}
