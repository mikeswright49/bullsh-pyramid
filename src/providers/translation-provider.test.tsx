import React, { useContext } from 'react';
import { TranslationContext, TranslationProvider } from './translation-provider';
import { render } from '@testing-library/react';
import { MOCK_GAME_STATE } from 'src/__tests__/data/mock-game-state';
import { GameContext } from './game-provider';

describe('Translation Provider', () => {
    function ComponentMock() {
        const data = useContext(TranslationContext);
        return <div>{JSON.stringify(data)}</div>;
    }
    it('should be able to load and handle translations', () => {
        const { container } = render(
            <GameContext.Provider value={MOCK_GAME_STATE}>
                <TranslationProvider>
                    <ComponentMock />
                </TranslationProvider>
            </GameContext.Provider>
        );
        expect(container).toMatchSnapshot();
    });
});
