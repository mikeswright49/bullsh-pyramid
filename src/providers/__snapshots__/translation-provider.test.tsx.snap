// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Translation Provider should be able to load and handle translations 1`] = `
<div>
  <div>
    {"translations":{"game.board.current.row":"Current Row: {{ activeRow }}","game.board.current.card":"Current card","game.display.stage.memorization.text":"You have 30s to remember your cards!","game.display.stage.flipping.text":"The host is flipping a card","game.display.stage.declaration.text":"Let the world know if you have it","game.display.stage.assign.text":"Players are beginning to lie left right and center","game.display.stage.bullshit.text":"People should know who you think was lieing","game.display.stage.reveal.text":"We all now know the truth","game.display.stage.memory.text":"Well which cards did you have?","game.display.stage.complete.text":"Well time for a new game","host.display.stage.initiation.title":"You're the host","host.display.stage.initiation.button":"Start game","host.display.stage.memorization.title":"You're still the host","host.display.stage.memorization.subtitle":"Everone has 30s to remember their cards!","host.display.stage.flipping.title":"Time to start flipping cards","host.display.stage.flipping.button":"Flip card","host.display.stage.declaration.title":"These players have said they have it","host.display.stage.declaration.button":"Start giving out drinks","host.display.stage.assign.title":"Players are assigning points out now","host.display.stage.assign.button":"Find who has to do what","host.display.stage.bullshit.title":"Click the button to find out whats next","host.display.stage.bullshit.button":"Reveal the truth","host.display.stage.reveal.title":"Move onto next round","host.display.stage.reveal.button":"Flip card","vote.assign.title":"Assign {{count}} \\"points\\" out","vote.assign.submit":"All done","vote.assign.complete":"Well that was easy","global.player_name":"Player name","global.points":"points"}}
  </div>
</div>
`;
